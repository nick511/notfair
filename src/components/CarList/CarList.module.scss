
.component {
  padding: 30px 0;
}
