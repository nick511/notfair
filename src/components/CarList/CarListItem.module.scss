@import "style/variables.scss";
@import "style/helpers.scss";


.component {
  display: flex;
  @include clearfix;
  padding: 15px 25px 0;
  margin-bottom: 2px;
  background: white;

  @include breakpoint(tablet, max) {
    margin-right: 20px;
    margin-left: 20px;
  }
  @include breakpoint(tablet-small, max) {
    flex-direction: column;
  }
}

.image {
  position: relative;
  overflow: hidden;
  width: 250px;
  height: 160px;

  @include breakpoint(tablet-small, max) {
    width: 100%;
    height: 200px;
  }

  img {
    position: absolute;
    left: 50%;
    top: 50%;
    width: 100%;
    height: auto;
    transform: translate(-50%,-50%);
  }
}

.info {
  position: relative;
  width: 300px;
  padding: 25px;
  line-height: 1.3em;
  text-align-last: left;
  color: $light-gray;

  @include breakpoint(tablet-small, max) {
    width: 100%;
    padding-left: 0;
    padding-right: 0;
    font-size: $mobile-body-size;
  }

  .name {
    margin-bottom: 10px;
    color: $gray;
    font-size: $label-size;
  }

  .price {
    color: $brand-color;
    font-size: 2em;
    margin-right: 5px;
  }

  .infoItem {
    float: left;
    width: 50%;
    line-height: 1.5em;
  }

  .favorite {
    position: absolute;
    top: 10px;
    right: 10px;

    @include breakpoint(tablet-wide) {
      top: 20px;
      right: 10px;
    }
    @include breakpoint(phablet, max) {
      top: 20px;
      right: 0px;
    }
  }
}

.priceChart {
  flex: 1;

  @include breakpoint(tablet-wide, max) {
    display: none;
  }
}
