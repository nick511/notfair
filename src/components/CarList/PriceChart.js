import React from 'react'
import PropTypes from 'prop-types'

import styles from './PriceChart.module.scss'

function PriceChart({ realPrice, minPrice, listPrice, fairMaxPrice }) {
  const getPricePercentage = price => {
    let pcercent = ((price - minPrice) / (listPrice - minPrice)) * 100
    return Math.min(Math.max(parseInt(pcercent), 0), 100) // limit number between 0-100
  }

  const formatCurrency = number => '$' + parseInt(number, 10).toLocaleString()

  const realPricePercent = getPricePercentage(realPrice)
  const fairMaxPercent = getPricePercentage(fairMaxPrice)

  return (
    <div className={styles.component}>
      <div className={styles.minPrice}>
        <div className={styles.label}>Min Price</div>
        <div className={styles.price}>{formatCurrency(minPrice)}</div>
      </div>

      <div className={styles.priceBar}>
        <div className={styles.barBox}>
          <div className={styles.bar}>
            <div
              className={styles.colorBar}
              style={{ width: `${fairMaxPercent}%` }}
            />
          </div>
          <div
            className={styles.realPrice}
            style={{ left: `${realPricePercent}%` }}
          >
            <div className={styles.price}>{formatCurrency(realPrice)}</div>
            <div className={styles.line} />
            <div className={styles.roundButton} />
          </div>
          <div
            className={styles.fairMax}
            style={{ left: `${fairMaxPercent}%` }}
          >
            <div className={styles.line} />
            <div className={styles.offset}>
              <div className={styles.label}>Fair Program Max</div>
              <div className={styles.price}>{formatCurrency(fairMaxPrice)}</div>
            </div>
          </div>
        </div>
      </div>

      <div className={styles.listPrice}>
        <div className={styles.label}>List Price</div>
        <div className={styles.price}>{formatCurrency(listPrice)}</div>
      </div>
    </div>
  )
}

PriceChart.propTypes = {
  realPrice: PropTypes.number.isRequired,
  minPrice: PropTypes.number.isRequired,
  listPrice: PropTypes.number.isRequired,
  fairMaxPrice: PropTypes.number.isRequired,
}

export default PriceChart
