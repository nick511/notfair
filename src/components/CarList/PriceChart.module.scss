@import "style/variables.scss";

$price-line-height: 60px;

.component {
  display: flex;
  align-items: center;
  margin-top: 40px;
  line-height: 1.4em;
}

.label {
  font-size: $label-size;
}

.price {
  color: $light-gray;
}

.minPrice,
.listPrice {
  display: flex;
  flex-direction: column;
  justify-content: center;
  width: 85px;
  height: $price-line-height;
}

.minPrice {
  padding-right: 20px;
  border-right: 1px solid $line-v-color;
  text-align: right;
}

.listPrice {
  padding-left: 20px;
  border-left: 1px solid $line-v-color;
  text-align: left;
}

.priceBar {
  flex: 1;

  .barBox {
    position: relative;
    margin: auto;
    width: 90%;
    $bar-height: 10px;
    height: $bar-height;

    .bar {
      border-radius: $bar-height/2;
      background-color: $light-gray;
      overflow: hidden;

      .colorBar {
        height: $bar-height;
        background-color: $brand-color;
      }
    }

    .realPrice {
      position: absolute;
      bottom: -12px;
      text-align: center;

      .price,
      .roundButton {
        transform: translate(-50%);
      }

      .price {
        color: inherit;
      }

      .line {
        width: 1px;
        height: 12px;
        background-color: $brand-color;
        margin-top: 5px;
      }

      .roundButton {
        width: 35px;
        height: 35px;
        background-color: #EDEDED;
        border-radius: 100%;
        border: 1px solid #bbb;
        box-shadow: 1px 1px 1px rgba(0, 0, 0, 0.2);
      }
    }

    .fairMax {
      position: absolute;
      top: -($price-line-height - $bar-height)/2;
      text-align: center;

      .line {
        width: 1px;
        height: $price-line-height;
        background-color: $line-v-color;
      }

      .offset {
        transform: translate(-50%);
        white-space: nowrap;
      }
    }

  }
}
