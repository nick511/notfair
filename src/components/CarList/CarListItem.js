import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'
import { Link } from 'react-router-dom'
import cn from 'classnames'

import PriceChart from './PriceChart'
import styles from './CarListItem.module.scss'

class CarListItem extends PureComponent {
  _favoriteCar = e => {
    e.preventDefault()
    const { favoriteCarAction, car } = this.props
    favoriteCarAction(car.id, !car.isFavorite)
  }

  render() {
    const { car } = this.props

    const payment = Array.isArray(car.product_financials)
      ? car.product_financials[0]
      : null

    return (
      <Link to={`/vehicles/${car.id}`} className={styles.component}>
        {/* Car image */}
        <div className={styles.image}>
          <img src={car.chrome_image_url} alt="" />
        </div>

        {/* Car info */}
        <div className={styles.info}>
          <div className={styles.name}>
            {car.model_year} {car.make} {car.model}
          </div>

          {payment && (
            <div>
              <span className={styles.price}>
                ${payment.monthly_payment_cents / 100}
              </span>
              per mo.
            </div>
          )}

          <div className={styles.infoItem}>TRIM: {car.trim}</div>
          <div className={styles.infoItem}>MILES: {car.mileage}</div>

          {payment && (
            <div className={styles.infoItem}>
              Start Fee: ${payment.start_fee_cents / 100}
            </div>
          )}

          {/* Favorite button */}
          <div
            className={cn(styles.favorite, 'favorite', {
              favorited: car.isFavorite,
            })}
            onClick={this._favoriteCar}
          >
            <i className="fa fa-heart" />
          </div>
        </div>

        <div className={styles.priceChart}>
          <PriceChart
            realPrice={16999}
            minPrice={12680}
            listPrice={17999}
            fairMaxPrice={17750}
          />
        </div>
      </Link>
    )
  }
}

CarListItem.propTypes = {
  car: PropTypes.shape({
    id: PropTypes.string.isRequired,
    make: PropTypes.string.isRequired,
    model: PropTypes.string.isRequired,
  }).isRequired,
  favoriteCarAction: PropTypes.func.isRequired,
}

export default CarListItem
