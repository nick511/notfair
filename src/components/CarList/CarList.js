import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'

import ErrorBoundary from 'components/Utils/ErrorBoundary'
import CarListItem from 'components/CarList/CarListItem'
import styles from './CarList.module.scss'

class CarList extends PureComponent {
  render() {
    const { cars, favoriteCarAction } = this.props
    return (
      <div className={styles.component}>
        {Object.keys(cars).map(key => (
          <ErrorBoundary key={key}>
            <CarListItem
              car={cars[key]}
              favoriteCarAction={favoriteCarAction}
            />
          </ErrorBoundary>
        ))}
      </div>
    )
  }
}

CarList.propTypes = {
  cars: PropTypes.object.isRequired,
  favoriteCarAction: PropTypes.func.isRequired,
}

export default CarList
