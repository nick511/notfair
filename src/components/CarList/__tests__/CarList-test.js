import React from 'react'
import { shallow } from 'enzyme'

import CarList from '../CarList'

describe('<CarList />', () => {
  let wrapper = null
  const cars = {
    allIds: ['19XFC2F59GE2276732016'],
    byId: {
      '19XFC2F59GE2276732016': {
        id: '19XFC2F59GE2276732016',
        make: 'Honda',
        mileage: 35292,
        model: 'Civic Sedan',
      },
    },
  }
  const hasMore = false
  const fetchCars = jest.fn()
  const favoriteCarAction = jest.fn()

  beforeEach(() => {
    wrapper = shallow(
      <CarList
        cars={cars}
        hasMore={hasMore}
        fetchCars={fetchCars}
        favoriteCarAction={favoriteCarAction}
      />,
    )
  })

  it('renders without crashing', () => {
    expect(wrapper).toMatchSnapshot()
  })

  it('renders CarListItem', () => {
    expect(wrapper.find('CarListItem')).toHaveLength(1)
  })
})
