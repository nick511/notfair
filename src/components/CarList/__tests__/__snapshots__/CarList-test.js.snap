// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CarList /> renders without crashing 1`] = `
<div
  className="component"
>
  <InfiniteScroll
    element="div"
    getScrollParent={null}
    hasMore={false}
    initialLoad={true}
    isReverse={false}
    loadMore={[MockFunction]}
    loader={
      <div
        className="loader"
      >
        Loading ...
      </div>
    }
    pageStart={0}
    ref={null}
    threshold={250}
    useCapture={false}
    useWindow={true}
  >
    <ErrorBoundary
      key="19XFC2F59GE22767320160"
    >
      <CarListItem
        car={
          Object {
            "id": "19XFC2F59GE2276732016",
            "make": "Honda",
            "mileage": 35292,
            "model": "Civic Sedan",
          }
        }
        favoriteCarAction={[MockFunction]}
      />
    </ErrorBoundary>
  </InfiniteScroll>
</div>
`;
